### PR TITLE
Notification Subsystem

### PR DESCRIPTION
**Pull request by cnangeh**
Work done:

- Work done by Adhyan last year
- User subscription to topic on user creation
- Topic creation on entity creation
- Upgrade process of user notification

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/notification/domain/Topic.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+import org.apache.fineract.infrastructure.core.api.JsonCommand;
+import org.apache.fineract.infrastructure.core.domain.AbstractPersistableCustom;
+
+@Entity
+@Table(name = "topic")
+public class Topic extends AbstractPersistableCustom<Long> {
+	
+	
+	@Column(name = "title", unique = true, nullable = false, length = 100)
+    private String title;
```
Comment (pranjal-goswami) [inline fineract-provider/src/main/java/org/apache/fineract/notification/domain/Topic.java]: Indentation is inconsistent. Check other attributes also for indentation below.

File: fineract-provider/src/main/java/org/apache/fineract/notification/domain/Topic.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+	public static Topic fromJson(final JsonCommand command) {
+
+		String title = "";
+		Boolean enabled = null;
+		Long entityId = 0L;
+		String entityType = "";
+		String memberType = "";
+		
+		if(command.hasParameter("title")) {
```
Comment (pranjal-goswami) [inline fineract-provider/src/main/java/org/apache/fineract/notification/domain/Topic.java]: Inconsistent indentation

File: fineract-provider/src/main/java/org/apache/fineract/notification/eventandlistener/NotificationEventListener.java [hunk 3]
```diff
@@ -61,17 +61,7 @@ public void onMessage(Message message, Session session) throws JMSException {
...
-					AppUser appUser = appUserRepository.findOne(userId);
-					if (!Objects.equals(appUser.getOffice().getId(), notificationData.getOfficeId())) {
-						userIds.remove(userId);
-					}
-				}
-			}
-			
+						
 			if (userIds.contains(appUserId)) {
 				userIds.remove(appUserId);
```
Comment (pranjal-goswami) [inline fineract-provider/src/main/java/org/apache/fineract/notification/eventandlistener/NotificationEventListener.java]: What exactly is this line doing? or what exactly is the method achieving?

File: fineract-provider/src/main/java/org/apache/fineract/notification/service/NotificationDomainServiceImpl.java [hunk 4]
```diff
@@ -552,7 +557,7 @@ private void buildNotification(String permission, String objectType, Long object
 		
 		String tenantIdentifier = ThreadLocalContextUtil.getTenant().getTenantIdentifier();
 		Queue queue = new ActiveMQQueue("NotificationQueue");
-		List<Long> userIds = retrieveUsersWithSpecificPermission(permission);
+		List<Long> userIds = retrieveSubscribers(officeId, permission);
```
Comment (pranjal-goswami) [inline fineract-provider/src/main/java/org/apache/fineract/notification/service/NotificationDomainServiceImpl.java]: At the time of deployment of this code, we can assume that there will be no topics that have been created for existing offices and users. What is your plan to populate the topics table for existing offices and users?

**Review by pranjal-goswami — CHANGES_REQUESTED** [review]

File: fineract-provider/src/main/java/org/apache/fineract/notification/eventandlistener/NotificationEventListener.java (hunk 3 above)
Comment (cnangeh) [reply fineract-provider/src/main/java/org/apache/fineract/notification/eventandlistener/NotificationEventListener.java]: If prevents the user who generated this notification from being notified.

File: fineract-provider/src/main/java/org/apache/fineract/notification/service/NotificationDomainServiceImpl.java (hunk 4 above)
Comment (cnangeh) [reply fineract-provider/src/main/java/org/apache/fineract/notification/service/NotificationDomainServiceImpl.java]: I will write a migration script that will take care of that.

**Pull request closed without merging**